### PR TITLE
Refactor test state related code

### PR DESCRIPTION
**Pull request by kasperski95**
Closes: https://github.com/software-mansion/protostar/issues/1389

### CODE REVIEW

File: protostar/starknet/__init__.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+from starkware.starknet.business_logic.state.state import CachedState as StarknetState
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply protostar/starknet/__init__.py:1]: Moreover it may be confused with `starkware.starknet.testing.state.StarknetState` which is a totally different thing

File: protostar/cheatable_starknet/cheatables/cheatable_syscall_handler.py [hunk 2]
```diff
@@ -31,31 +41,68 @@ def __init__(self, message: str):
 
 
 class CheatableSysCallHandler(BusinessLogicSysCallHandler):
-    def __init__(self, state: StateSyncifier, **kwargs: Any):
-        self.cheatable_state: "CheatableCachedState" = cast(
-            "CheatableCachedState", state.async_state
-        )
+    block_info_controller_state: Optional["BlockInfoControllerState"] = None
+    contracts_controller_state: Optional["ContractsControllerState"] = None
```
Comment (Arcticae) [inline protostar/cheatable_starknet/cheatables/cheatable_syscall_handler.py:45]: This is not a dataclass, which makes those class variables

File: protostar/cairo_testing/cairo_test_execution_state.py [hunk 3]
```diff
@@ -10,45 +11,164 @@
...
+        self._emitted_events.extend(emitted_events)
+
+    def set_pranked_address(self, target_address: Address, pranked_address: Address):
+        self._target_address_to_pranked_address[target_address] = pranked_address
+
+    def remove_pranked_address(self, target_address: Address):
+        if target_address in self._target_address_to_pranked_address:
+            del self._target_address_to_pranked_address[target_address]
+
+    def clone(self):
```
Comment (Arcticae) [inline protostar/cairo_testing/cairo_test_execution_state.py:98]: The naming is not consistent, we have `fork`, `copy`, and now `clone`. I don't see a particular difference between those.

File: protostar/cairo_testing/cairo_test_execution_state.py [hunk 4]
```diff
@@ -10,45 +11,164 @@
...
+    def get_class_hash_from_address(self, contract_address: Address) -> ClassHashType:
+        return self._contract_address_to_class_hash[contract_address]
+
+    def get_contract_abi_from_class_hash(self, class_hash: ClassHashType) -> AbiType:
+        return self._class_hash_to_contract_abi[class_hash]
+
+    def get_contract_abi_from_contract_address(
+        self, contract_address: Address
+    ) -> AbiType:
+        class_hash = self._contract_address_to_class_hash[contract_address]
```
Comment (Arcticae) [inline protostar/cairo_testing/cairo_test_execution_state.py:64]: This can use `self.get_class_hash_from_address`

File: protostar/cairo_testing/execution_environments/cairo_execution_environment.py [hunk 5]
```diff
@@ -28,6 +31,12 @@ async def run_cairo_function(
         *args: Any,
         **kwargs: Any,
     ):
+        CheatableSysCallHandler.block_info_controller_state = (
+            self.state.block_info_controller_state
+        )
+        CheatableSysCallHandler.contracts_controller_state = (
+            self.state.contracts_controller_state
+        )
```
Comment (Arcticae) [inline protostar/cairo_testing/execution_environments/cairo_execution_environment.py:39]: Setting a static class property with a state instance is a big no-no, it was implemented before and was marked for removal in future versions. We'd like to avoid this.

**Review by Arcticae — CHANGES_REQUESTED** [review]
I like the separation of concerns in the state, although `starknet_state` and `state` are not the names i agree with, it might be correct from user standpoint but not from a technical one. 

I also feel like the states & controllers are now tightly coupled and could be merged into one. I don't see a case where we need a state without a controller right now, or vice-versa. Please correct me if i'm wrong though.

File: protostar/starknet/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet/__init__.py:1]: `CachedState` doesn't make much sense from Protostar perspective.

[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/starknet/__init__.py:1]: Even though it is the case, it is still hard to follow. It is better to use one name for one thing, even though it is imperfect. It is easier to know what we are talking about. Something like this will cause a lot of clarification questions in tech discussions "Do we talk about renamed cached state?"

File: protostar/cheatable_starknet/cheatables/cheatable_syscall_handler.py (hunk 2 above)
Comment (kasperski95) [reply protostar/cheatable_starknet/cheatables/cheatable_syscall_handler.py:45]: Resolving in favor of https://github.com/software-mansion/protostar/pull/1451#discussion_r1101478245

**Comment by kasperski95** [comment]
> I also feel like the states & controllers are now tightly coupled and could be merged into one. I don't see a case where we need a state without a controller right now, or vice-versa. Please correct me if i'm wrong though.

`ExpectEventsController` depends on `ContractsControllerState`. ContractsController is already big and messy. It acts as a Controller (Business logic) and an Interfacer (to Starknet). Adding another concern will create more mess. State need to be forkable.

Instead of joining controllers and state, I think I would be better to join `ContractsContractState` and `BlockInfoContractState` into one.

File: protostar/cairo_testing/execution_environments/cairo_execution_environment.py (hunk 5 above)
Comment (MaksymilianDemitraszek) [reply protostar/cairo_testing/execution_environments/cairo_execution_environment.py:39]: Agreed, there is no obvious reason to do that, though there are some subtle, but important ones. First, it is much easier to follow and explore the codebase without monkey patching, it is always somewhat confisuing to reason about state, you can't never be sure if there is no race condition. Second reason is we are close to rust migration, and Rust does not allow such hacks. I would prefer to not encourage any form of dependence on them, to simplify the integration in the future. Hope it makes sense.

File: protostar/cairo_testing/cairo_test_execution_state.py [hunk 6]
```diff
@@ -10,45 +11,164 @@
...
         return dataclasses.replace(
             self,
             context=deepcopy(self.context),
             config=deepcopy(self.config),
             output_recorder=self.output_recorder.fork(),
             stopwatch=self.stopwatch.fork(),
-            starknet=self.starknet.copy(),
+            # pylint: disable=protected-access
+            starknet_state=self.starknet_state._copy(),
+            contracts_controller_state=self.contracts_controller_state.clone(),
```
Comment (MaksymilianDemitraszek) [inline protostar/cairo_testing/cairo_test_execution_state.py:170]: I believe it does not work properly. In the new architecture single test can execute multiple transactions, but you clone it only per test, not per transaction. We need copy apply mechanism implemented, thats why we used cached state before. Right now if you invoke transaction which reverts, protostar state may contain some lefovers from this transaciton. Does it make sense?

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]

File: protostar/cairo_testing/cairo_test_execution_state.py (hunk 6 above)
Comment (kasperski95) [reply protostar/cairo_testing/cairo_test_execution_state.py:170]: It would make sense, if Protostar's hint local were available in a contract. However, they are not. 
That means contract execution doesn't mutate Protostar state. Changes to Protostar state are done after contract execution. Therefore, we don't need to implement the copy and apply mechanism.

```py
with self._starknet_state.copy_and_apply() as state_copy:
            call_info = await entry_point.execute_for_testing(
                state=state_copy,
                general_config=StarknetGeneralConfig(),
            )
            self._add_emitted_events(call_info.get_sorted_events()) # <= this won't execute in case of revert
```

File: protostar/starknet/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Arcticae) [reply protostar/starknet/__init__.py:1]: this is also the term that will continue to be used in rust implementation, so it will be easier to reason about the migration to rust, and `legacy <-> new codebase` relation when we get to it.

File: protostar/cairo_testing/cairo_test_execution_state.py (hunk 6 above)
Comment (MaksymilianDemitraszek) [reply protostar/cairo_testing/cairo_test_execution_state.py:170]: Yea, but we store there deployed contracts with syscalls as well
